File: modules/rust_workloads/src/string_churn.rs
//! O(n²) string concatenation: every round allocates a fresh string and
//! copies the whole accumulator — high transient allocation rate, tiny live
//! heap at any instant.

pub fn concat(pieces: usize, piece: &str) -> String {
    let mut acc = String::new();
    for _ in 0..pieces {
        acc = format!("{acc}{piece}");
    }
    acc
}

File: modules/python_workloads/src/python_workloads/string_churn.py
"""O(n²) string concatenation: every round allocates a fresh string and
copies the whole accumulator — high transient allocation rate, tiny live
heap at any instant. The f-string sidesteps CPython's in-place resize
optimisation for `acc += piece`, keeping the churn honest."""


def concat(pieces: int, piece: str) -> str:
    acc = ""
    for _ in range(pieces):
        acc = f"{acc}{piece}"
    return acc
